WordPress/Sniffs/WP/CapabilitiesSniff.php

```diff
@@ -0,0 +1,301 @@
+		'delete_posts'                => true,
+		'read'                        => true,
+		'create_sites'                => true,
+		'delete_sites'                => true,
+		'manage_network'              => true,
+		'manage_sites'                => true,
+		'manage_network_users'        => true,
+		'manage_network_plugins'      => true,
+		'manage_network_themes'       => true,
+		'manage_network_options'      => true,
+		'delete_site'                 => true,
+		'upgrade_network'             => true,
+		'setup_network'               => true,
+		'upload_plugins'              => true,
+		'upload_themes'               => true,
+		'customize'                   => true,
+		'add_users'                   => true,
+		'install_languages'           => true,
+		'update_languages'            => true,
+		'deactivate_plugins'          => true,
+		'upgrade_php'                 => true,
+		'export_others_personal_data' => true,
+		'erase_others_personal_data'  => true,
+		'manage_privacy_options'      => true,
+		'edit_categories'             => true,
+		'delete_categories'           => true,
+		'manage_post_tags'            => true,
+		'edit_post_tags'              => true,
+		'delete_post_tags'            => true,
+		'edit_css'                    => true,
+		'assign_categories'           => true,
+		'assign_post_tags'            => true,
+	);
+
+	/**
+	 * List of deprecated core capabilities.
+	 *
+	 * User Levels were  deprecated in version 3.0.
+	 * To be updated after every major release.
+	 * Last updated for WordPress 4.9.6.
+	 *
+	 * @link https://github.com/WordPress/wordpress-develop/blob/master/tests/phpunit/tests/user/capabilities.php
+	 *
+	 * @since 1.0.0
+	 *
+	 * @var array All deprecated capabilities in core.
+	 */
+	protected $deprecated_capabilities = array(
+		'level_10' => true,
+		'level_9'  => true,
+		'level_8'  => true,
+		'level_7'  => true,
+		'level_6'  => true,
+		'level_5'  => true,
+		'level_4'  => true,
+		'level_3'  => true,
+		'level_2'  => true,
+		'level_1'  => true,
+		'level_0'  => true,
+	);
+
+	/**
+	 * Process the parameters of a matched function.
+	 *
+	 * @since 1.0.0
+	 *
+	 * @param int    $stackPtr        The position of the current token in the stack.
+	 * @param array  $group_name      The name of the group which was matched.
+	 * @param string $matched_content The token content (function name) which was matched.
+	 * @param array  $parameters      Array with information about the parameters.
+	 *
+	 * @return void
+	 */
+	public function process_parameters( $stackPtr, $group_name, $matched_content, $parameters ) {
+
+		$position = $this->target_functions[ $matched_content ];
+		if ( ! isset( $parameters[ $position ] ) ) {
+			return;
+		}
+
+		$next_not_empty = $this->phpcsFile->findNext(
+			Tokens::$emptyTokens,
+			$parameters[ $position ]['start'],
+			$parameters[ $position ]['end'] + 1,
+			true
+		);
+
+		$matched_parameter = $this->strip_quotes( $this->tokens[ $next_not_empty ]['content'] );
```

Comment:
Token walking means doing a `for()` loop between the param start and end. You are now just checking the first token and not even checking that this is a text string token....

This will break on all of the following examples (silly code, but you get my point):
```php
if ( author_can( $post, editor() ) ) { } // Should not error for using role.
if ( author_can( $post, 'super' . '_admin' ) ) { } // Should give error for using role.

// @codingStandardsChangeSetting WordPress.WP.Capabilities check_only_known_caps false
if ( author_can( $post, editor() ) ) { } // Should give warning about unknown capability, not error for using role.
if ( author_can( $post, install_plugins() ) ) { } // Should give warning about unknown capability.
if ( author_can( $post, 'install_' . 'plugins' ) ) { } // OK.
if ( author_can( $post, 'install_plugins' . $something ) ) { } // Should give warning about unknown capability.
if ( author_can( $post, 'install_plugins' . '_for_theme' ) ) { } // Should give warning about unknown capability.
// @codingStandardsChangeSetting WordPress.WP.Capabilities check_only_known_caps true
```

Reply:
Would it make sense to have this code as a utility function? I think there are other classes that would profit from it. 

```php
// More extensive check to prevent people circumventing the sniff.
$text                 = '';
$found_variable_token = false;

for ( $i = $parameters[1]['start']; $i <= $parameters[1]['end']; $i++ ) {
    if ( isset( Tokens::$stringTokens[ $this->tokens[ $i ]['code'] ] ) === false ) {
        if ( T_VARIABLE === $this->tokens[ $i ]['code'] ) {
            $found_variable_token = true;
        }

        return;
    }

    $text .= $this->strip_quotes( $this->tokens[ $i ]['content'] );
}
```
Copied the code from https://github.com/WPTRT/WordPress-Coding-Standards/pull/26/

Reply:
I was thinking about that, but I don't think there has been much need for it before now in WPCS.

All the same, I'd be happy for this to become a utility function, however, the "is there a variable ?" check in this code snippet would probably be a different question whenever used. Also in the above sample, it is presumed only single/double quoted strings are used, not heredoc/nowdoc, so some changes would be needed and I'd guess that the utility should in that case just retrieve a string of all text string tokens found between `$start` and `$end`, not do the "is there are variable" check ?

P.S.: looking at the code now, I'm still going to make one more change to the WPTRT sniff this came from ;-)

Reply:
OH, I forgot that the code was checking for two things. When I mentioned the idea I was thinking about returning a string of all of the text string token.